File: app/src/main/java/com/igordanilchik/daggertest/DaggerApplication.java
package com.igordanilchik.daggertest;

import android.app.Application;
import android.content.Context;

import com.squareup.leakcanary.LeakCanary;

public class DaggerApplication extends Application {

    public static DaggerApplication get(Context context) {
        return (DaggerApplication) context.getApplicationContext();
    }

    private AppComponent appComponent;
    private RepositoryComponent repositoryComponent;

    @Override
    public void onCreate() {
        super.onCreate();
        initInjector();
        initLeakDetection();
    }

    private void initLeakDetection() {
        if (LeakCanary.isInAnalyzerProcess(this)) {
            // This process is dedicated to LeakCanary for heap analysis.
            // You should not init your app in this process.
            return;
        }
        if (BuildConfig.DEBUG) {
            LeakCanary.install(this);

//            StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
//                        .detectAll()
//                        .penaltyLog()
//                        .build());
//            StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
//                        .detectAll()
//                        .penaltyLog()
//                        //.penaltyDeath()
//                        .build());
        }
    }

    private void initInjector() {
        appComponent = DaggerAppComponent.builder()
                .appModule(new AppModule(this))
                .httpClientModule(new HttpClientModule())
                .build();
    }

    public AppComponent getAppComponent() {
        return appComponent;
    }

    public RepositoryComponent getRepositoryComponent() {
        if (repositoryComponent == null) {
            repositoryComponent = appComponent.plus(new RepositoryModule());
        }
        return repositoryComponent;
    }
}

File: app/src/main/java/com/igordanilchik/daggertest/app/DaggerApplication.kt
package com.igordanilchik.daggertest.app

import android.app.Application
import android.content.Context
import com.igordanilchik.daggertest.BuildConfig
import com.igordanilchik.daggertest.common.di.ApplicationComponent
import com.igordanilchik.daggertest.common.di.ApplicationModule
import com.igordanilchik.daggertest.common.di.DaggerApplicationComponent
import com.igordanilchik.daggertest.common.di.RepositoryModule
import com.igordanilchik.daggertest.common.log.ReleaseTree
import com.squareup.leakcanary.LeakCanary
import timber.log.Timber

class DaggerApplication : Application() {

    lateinit var appComponent: ApplicationComponent
        private set

    override fun onCreate() {
        super.onCreate()
        initInjector()
        initLeakDetection()
        initTimber()
    }

    private fun initLeakDetection() {
        if (LeakCanary.isInAnalyzerProcess(this)) {
            // This process is dedicated to LeakCanary for heap analysis.
            // You should not init your app in this process.
            return
        }
        if (BuildConfig.DEBUG) {
            LeakCanary.install(this)

//            StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.Builder()
//                    .detectAll()
//                    .penaltyLog()
//                    .build())
//            StrictMode.setVmPolicy(StrictMode.VmPolicy.Builder()
//                    .detectAll()
//                    .penaltyLog()
//                    //.penaltyDeath()
//                    .build())
        }
    }

    private fun initInjector() {
        appComponent = DaggerApplicationComponent.builder()
                .applicationModule(ApplicationModule(this))
                .repositoryModule(RepositoryModule())
                .build()
    }

    /**
     * Plant proper Timber for Logging for Debug
     * or Sending Errors as Crashes for Release
     * Pretends as Fail!
     */
    private fun initTimber() {
        if (BuildConfig.DEBUG) {
            Timber.plant(Timber.DebugTree())
        } else {
            Timber.plant(ReleaseTree())
        }
    }

    companion object {

        operator fun get(context: Context): DaggerApplication {
            return context.applicationContext as DaggerApplication
        }
    }

}
